deploy/template.yaml

```diff
@@ -299,7 +304,171 @@ Resources:
           ToPort: 443
       VpcId:
         Fn::ImportValue: !Sub "${VpcStackName}-VpcId"
-  
+
+  PartialNameMatchBucketKey:
+    Type: AWS::KMS::Key
+    Properties:
+      Description: Specific KMS CMK for Partial Matches S3 Bucket Encryption
+      Enabled: true
+      KeyPolicy:
+        Version: "2012-10-17"
+        Statement:
+          - Sid: AllowIdentityPoliciesInAccountToDetermineKeyAccess
+            Effect: Allow
+            Principal:
+              AWS: !Sub "arn:aws:iam::${AWS::AccountId}:root"
+            Action:
+              - "kms:*"
+            Resource:
+              - "*"
+      KeySpec: SYMMETRIC_DEFAULT
+      KeyUsage: ENCRYPT_DECRYPT
+      MultiRegion: false
+      PendingWindowInDays: 7
+      Tags:
+        - Key: KeyType
+          Value: Encryption Key
+        - Key: Environment
+          Value: !Sub ${Environment}
+
+  PartialNameMatchBucketKeyAlias:
+    Type: AWS::KMS::Alias
+    Properties:
+      AliasName: !If
+        -  CreateDevResources
+        - !Sub "alias/PartialNameMatchBucketEncryptionKey-${AWS::StackName}"
+        - alias/PartialNameMatchBucketEncryptionKey
+      TargetKeyId: !Ref PartialNameMatchBucketKey
+
+  PartialNameMatchBucketPolicy:
+    Type: AWS::S3::BucketPolicy
+    Properties:
+      Bucket: !Ref PartialNameMatchBucket
+      PolicyDocument:
+        Version: "2012-10-17"
+        Statement:
+          - Sid: DenyIncorrectEncryptionType
+            Effect: Deny
+            Principal: "*"
+            Action:
+              - s3:PutObject
+            Resource:
+              - !Join ["", [ "arn:aws:s3:::", !Ref PartialNameMatchBucket, "/*"]]
+            Condition:
+              StringNotEquals:
+                "s3:x-amz-server-side-encryption": "aws:kms"
+          - Sid: DenyUnencryptedObjectUploads
+            Effect: Deny
+            Principal: "*"
+            Action:
+              - s3:PutObject
+            Resource:
+              - !Join ["", [ "arn:aws:s3:::", !Ref PartialNameMatchBucket, "/*"]]
+            Condition:
+              "Null":
+                "s3:x-amz-server-side-encryption": true
+          - Sid: RestrictGetObjectToAthenaOnly
+            Effect: Deny
+            Principal: "*"
+            Action:
+              - "s3:GetObject*"
+            Resource:
+              - !Join ["", [ "arn:aws:s3:::", !Ref PartialNameMatchBucket, "/*"]]
+            Condition:
+              StringNotEquals:
+                "aws:CalledVia": "athena.amazonaws.com"
+    DependsOn: PartialNameMatchBucket
+
+  PartialNameMatchBucket:
+    Type: AWS::S3::Bucket
+    DeletionPolicy: Delete
+    Properties:
+      BucketName: !Sub ${AWS::StackName}-partial-name-match-${Environment}
```

Comment:
I'm conscious that s3 buckets make stack deletion annoying. As this isn't required for core functionality do you think we could add a statement so that this is not deployed for custom stacks? (not necessarily now maybe as a fast follow)

Reply:
We need it for every stack not just the main ones?

Reply:
Can our custom stacks not push to the main s3 bucket?

Reply:
They can but then Athena will pull data from multiple stacks into single query and it might get confusing no?